# Part 1

In this part of the project, you will design and create a knowledge base, and then
compare different methods for performing inference on that knowledge base. <br>

**Description of the problem** <br>

- *Anjali*, *Brian*, and *Chen* graduated from QUB on different days this summer. <br>

- One of them studied *French*, one studied *Geography*, one read *History*. <br>

- One scored a *very good 65%*, one acheived an *excellent 70%*, and the other an *outstanding 75%*. <br>

- They graduated on the *2nd*, *3rd*, and *4th* July. <br>

We have acquired some knowledge about who studied what, what their marks were,
and when they graduated. However, our knowledge is incomplete. <br>


**Here is what we know:** <br>

1. Chen’s mark was lower than the mark of the person who studied French. <br>

2. The three people are Brian, the person who studied History, the person who graduated on the 2nd. <br>

3. Chen scored 70%, unless she studied history in which case her mark was 65%. <br>

4. The mark of the person who graduated on the 3rd was 5% higher than Chen’s. <br>

5. Anjali did not graduate on the 2nd. <br>

**The goal of this assignment is to determine who acheived which mark, what each person studied, and when they graduated** <br>

## Task (a)

Formulate a knowledge base (KB) for this problem using propositional logic. Explicitly state how each clue entails the sentences in the knowledge
base that are derived from it.

**Table of Available Knowledge** <br>

| Student | Subject | Mark | Graduated Date |
| ----------- | ----------- | ----------- | ----------- |
| Anjali | French | Very Good 65% | 2nd July |
| Brian | Geography | Excellent 70% | 3rd July |
| Chen | History | Outstanding 75% | 4th July |

 **Atomic Propositions** <br>

$A_F$: Anjali studied French. <br>
$B_F$: Brian studied French. <br>
$C_F$: Chen studied French. <br>

$A_G$: Anjali studied Geography. <br>
$B_G$: Brian studied Geography. <br>
$C_G$: Chen studied Geography. <br>

$A_H$: Anjali studied History. <br>
$B_H$: Brian studied History. <br>
$C_H$: Chen studied History. <br>

$A_{65}$: Anjali scored Very Good 65%. <br>
$B_{65}$: Brian scored Very Good 65%. <br>
$C_{65}$: Chen scored Very Good 65% <br>

$A_{70}$: Anjali scored Excellent 70%. <br>
$B_{70}$: Brian scored Excellent 70%. <br>
$C_{70}$: Chen scored Excellent 70%. <br>

$A_{75}$: Anjali scored Oustanding 75%. <br>
$B_{75}$: Brian scored Oustanding 75%. <br>
$C_{75}$: Chen scored Oustanding 75%. <br>

$A_{2}$: Anjali graduated on 2nd July. <br>
$B_{2}$: Brian graduated on 2nd July. <br>
$C_{2}$: Chen graduated on 2nd July. <br>

$A_{3}$: Anjali graduated on 3rd July. <br>
$B_{3}$: Brian graduated on 3rd July. <br>
$C_{3}$: Chen graduated on 3rd July. <br>

$A_{4}$: Anjali graduated on 4th July. <br>
$B_{4}$: Brian graduated on 4th July. <br>
$C_{4}$: Chen graduated on 4th July. <br>

**Clues**

**1. Chen’s mark was lower than the mark of the person who studied French.** <br>

Note[1]: Based on Clue(1), Chen didn't study French since there is the other student who studied French.

- $R_{1}$: $\lnot C_F$

Note[2]: If Chen didn't study French, then one of the other two students studied French.

- $R_{2}$: $\lnot C_F \implies (A_F \lor B_F)$

Note[3]: Based on Clue(1), Chen didn't score the highest mark in order for Chen's mark to be lower than the French student's mark.

- $R_{3}$: $\lnot C_{75}$

Note[4]: If Chen scored 70%, then Anjali studied French & scored 75% (or) Brian studied French & scored 75%.

- $R_{4}$: $C_{70} \implies (A_F \land A_{75}) \lor (B_F \land B_{75})$

Note[5]: If Chen scored 65%, then Anjali studied French & scored 75% (or) Brian studied French & scored 75% (or) Anjali studied French & scored 70% (or) Brian studied French & scored 70%.

- $R_{5}$: $C_{65} \implies (A_F \land A_{75}) \lor (B_F \land B_{75}) \lor (A_F \land A_{70}) \lor (B_F \land B_{70})  $

**2. The three people are Brian, the person who studied History, the person who graduated on the 2nd.** <br>

Note[6]: Based on Clue(2), Brian didn't study History since there is the other student who studied History.

- $R_{6}$: $\lnot B_H$

Note[7]: If Brian didn't study History, then one of the other two students studied History.

- $R_{7}$: $\lnot B_H \implies (A_H \lor C_H)$

Note[8 & 9]: Based on Clue(2), if a student studied History, then that student did not graduate on the 2nd.

- $R_{8}$: $A_H \implies \lnot A_{2}$

- $R_{9}$: $C_H \implies \lnot C_{2}$

Note[10]: Based on Clue(2), Brian didn't graduated on the 2nd since there is the other student who graduated on the 2nd.

- $R_{10}$: $\lnot B_{2}$

Note[11]: If Brian didn't graduate on the 2nd, then one of the other two students graduated on the 2nd.

- $R_{11}$: $\lnot B_{2} \implies (A_{2} \lor C_{2})$

Note[12 & 13]: Based on Clue(2), if a student graduated on the 2nd, then that student did not study History.

- $R_{12}$: $A_{2} \implies \lnot A_H$

- $R_{13}$: $C_{2} \implies \lnot C_H$

**3. Chen scored 70%, unless she studied history in which case her mark was 65%.** <br>

Note[14]: Based on Clue(3), if Chen didn't studied History, then Chen scored 70%.

- $R_{14}$: $\lnot C_H \implies C_{70}$

Note[15]: Based on Clue(3), if Chen studied History, then Chen scored 65%.

- $R_{15}$: $C_H \implies C_{65}$

**4. The mark of the person who graduated on the 3rd was 5% higher than Chen’s.** <br>

Note[16]: Based on Clue(4), Chen didn't graduate on the 3rd since there is the other student who graduated on the 3rd.

- $R_{16}$: $\lnot C_{3}$

Note[17]: If Chen didn't graduate on the 3rd, then one of the other two students graduated on the 3rd.

- $R_{17}$: $\lnot C_{3} \implies (A_{3} \lor B_{3})$

Note[18]: Based on Clue(4), if Chen got 70% then Anlaji graduated on the 3rd and scored 75% (or) Brian graduated on the 3rd and scored 75%

- $R_{18}$: $C_{70} \implies (A_{3} \land A_{75}) \lor (B_{3} \land B_{75})$

Note[19]: Based on Clue(4), If Chen got 65% then Anlaji graduated on the 3rd and scored 70% (or) Brian graduated on the 3rd and scored 70%

- $R_{19}$: $C_{65} \implies (A_{3} \land A_{70}) \lor (B_{3} \land B_{70})$


**5. Anjali did not graduate on the 2nd.** <br>

Note[20]: Based on Clue(5), Anjali did not graduate on the 2nd.

- $R_{20}$: $\lnot A_{2}$

Note[21]: If Anjali did not graduate on the 2nd, then one of the other two students graduated on the 2nd.

- $R_{21}$: $\lnot A_{2} \implies (B_{2} \lor C_{2})$

**Constraints** <br>

Note[22-39]: In order to avoid overlapping between students, subjects, scores & graduation dates, the following constraints will be applied.

The following sentences use the formula:
$f(P,Q,R) = (P \land\lnot Q\land\lnot R)\lor(\lnot P \land Q\land\lnot R)\lor(\lnot P \land\lnot Q\land R)$

- $R_{22}$: $f(A_F, A_G, A_H)$
- $R_{23}$: $f(B_F, B_G, B_H)$
- $R_{24}$: $f(C_F, C_G, C_H)$
<br>.
- $R_{25}$: $f(A_{65}, A_{70}, A_{75})$
- $R_{26}$: $f(B_{65}, B_{70}, B_{75})$
- $R_{27}$: $f(C_{65}, C_{70}, C_{75})$
<br>.
- $R_{28}$: $f(A_{2}, A_{3}, A_{4})$
- $R_{29}$: $f(B_{2}, B_{3}, B_{4})$
- $R_{30}$: $f(C_{2}, C_{3}, C_{4})$
<br>.
- $R_{31}$: $f(A_F, B_F, C_F)$
- $R_{32}$: $f(A_G, B_G, C_G)$
- $R_{33}$: $f(A_H, B_H, C_H)$
<br>.
- $R_{34}$: $f(A_{65}, B_{65}, C_{65})$
- $R_{35}$: $f(A_{70}, B_{70}, C_{70})$
- $R_{36}$: $f(A_{75}, B_{75}, C_{75})$
<br>.
- $R_{37}$: $f(A_{2}, B_{2}, C_{2})$
- $R_{38}$: $f(A_{3}, B_{3}, C_{3})$
- $R_{39}$: $f(A_{4}, B_{4}, C_{4})$

**Model**

|  0  |  1  |  2  |  3  |  4  |  5  |  6  |  7  |  8  |  9  | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 |
|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|-----|
|$A_F$|$B_F$|$C_F$|$A_G$|$B_G$|$C_G$|$A_H$|$B_H$|$C_H$|$A_{65}$|$B_{65}$|$C_{65}$|$A_{70}$|$B_{70}$|$C_{70}$|$A_{75}$|$B_{75}$|$C_{75}$|$A_{2}$|$B_{2}$|$C_{2}$|$A_{3}$|$B_{3}$|$C_{3}$|$A_{4}$|$B_{4}$|$C_{4}$|

In [ ]:
T = True
F = False

model =  [
    T, T, T,
    F, F, F,
    T, T, T,

    F, F, F,
    T, T, T,
    F, F, F,

    T, T, T,
    F, F, F,
    T, T, T
]

In [ ]:
implies = lambda P,Q: (not P) or Q

Constraint = lambda P,Q,R: (P and not Q and not R) or (not P and Q and not R) or (not P and not Q and R)

In [ ]:
KnowledgeBase = [
    # clue 1
    lambda model : not model[2], #R1
    lambda model : implies(not model[2], model[0] or model[1]), #R2

    lambda model : not model[17], # R3
    lambda model : implies(model[14], (model[0] and model[15]) or (model[1] and model[16])), # R4
    lambda model : implies(model[11], (model[0] and model[15]) or (model[1] and model[16]) or # R5
                                                                    (model[0] and model[12]) or (model[1] and model[13])),

    # clue 2
    lambda model : not model[7], # R6
    lambda model : implies(not model[7], model[6] or model[8]), # R7
    lambda model : implies(model[6], not model[18]), # R8
    lambda model : implies(model[8], not model[20]), # R9

    lambda model : not model[19], #R10
    lambda model : implies(not model[19], model[18] or model[20]), # R11
    lambda model : implies(model[18], not model[6]), # R12
    lambda model : implies(model[20], not model[8]), # R13


    # clue 3
    lambda model : implies(not model[8], model[14]), # R14
    lambda model : implies(model[8], model[11]), # R15

    # clue 4
    lambda model : not model[23], # R16
    lambda model : implies(not model[23], model[21] or model[22]), # R17
    lambda model : implies(model[11], (model[21] and model[12]) or (model[22] and model[13])), # R18
    lambda model : implies(model[14], (model[21] and model[15]) or (model[22] and model[16])), # R19

    # clue 5
    lambda model : not model[18], # R20
    lambda model : implies(not model[18], model[19] or model[20]), # R21

    # constraints
    lambda model : Constraint(model[0],model[3],model[6]), # R22
    lambda model : Constraint(model[1],model[4],model[7]), # R23
    lambda model : Constraint(model[2],model[5],model[8]), # R24

    lambda model : Constraint(model[9],model[12],model[15]), # R25
    lambda model : Constraint(model[10],model[13],model[16]), # R26
    lambda model : Constraint(model[11],model[14],model[17]), # R27

    lambda model : Constraint(model[18],model[21],model[24]), # R28
    lambda model : Constraint(model[19],model[22],model[25]), # R29
    lambda model : Constraint(model[20],model[23],model[26]), # R30

    lambda model : Constraint(model[0],model[1],model[2]), # R31
    lambda model : Constraint(model[3],model[4],model[5]), # R32
    lambda model : Constraint(model[6],model[7],model[8]), # R33

    lambda model : Constraint(model[9],model[10],model[11]), # R34
    lambda model : Constraint(model[12],model[13],model[14]), # R35
    lambda model : Constraint(model[15],model[16],model[17]), # R36

    lambda model : Constraint(model[18],model[19],model[20]), # R37
    lambda model : Constraint(model[21],model[22],model[23]), # R38
    lambda model : Constraint(model[24],model[25],model[26]), # R39
]

In [ ]:
import functools

def evaluateKB(model, KnowledgeBase):

    Result = [proposition(model) for proposition in KnowledgeBase]
    TrueOrFalse = functools.reduce(lambda a,b: a and b, Result)

    return TrueOrFalse, Result

In [ ]:
model =  [
    F, T, F,
    F, F, T,
    T, F, F,

    T, F, F,
    F, F, T,
    F, T, F,

    F, F, T,
    F, T, F,
    T, F, F
]

In [ ]:
TF, result = evaluateKB(model, KnowledgeBase)

print(f"{TF}: {result}")

True: [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]


## Task (b)


Use backtracking search to determine if your knowledge base is satisfiable and if so, identify a model that satisfies it.

In [ ]:
Clauses = [
    # clue 1
    ('~CF',), # R1
    ('CF', 'AF', 'BF'), # R2

    ('~C75',), # R3
    ('BF', 'AF', '~C70'), ('B75', 'AF', '~C70'), ('BF', 'A75', '~C70'), ('B75', 'A75', '~C70'), # R4
    ('BF', 'AF', '~C65'), ('B75', 'AF', 'BF', '~C65'), ('BF', 'A75', 'AF', '~C65'), ('B75', 'A75', 'AF', 'BF', '~C65'), ('BF', 'AF', 'A70', '~C65'), # R5
    ('B75', 'AF', 'A70', 'BF', '~C65'), ('BF', 'A75', 'A70', '~C65'), ('B75', 'A75', 'A70', 'BF', '~C65'), ('BF', 'AF', 'B70', '~C65'), ('B75', 'AF', 'B70', '~C65'),
    ('BF', 'A75', 'AF', 'B70', '~C65'), ('B75', 'A75', 'AF', 'B70', '~C65'), ('BF', 'AF', 'A70', 'B70', '~C65'), ('B75', 'AF', 'A70', 'B70', '~C65'),
    ('BF', 'A75', 'A70', 'B70', '~C65'),('B75', 'A75', 'A70', 'B70', '~C65'),

    # clue 2
    ('~BH',), # R6
    ('BH', 'AH', 'CH'), # R7
    ('~AH','~A2'), # R8
    ('~CH','~C2'), # R9

    ('~B2',), # R10
    ('B2', 'A2', 'C2'), # R11
    ('~A2','~AH'), # R12
    ('~C2','~CH'), # R13

    # clue 3
    ('CH', 'C70'), # R14
    ('~CH', 'C65'), # R15

    # clue 4
    ('~C3',), # R16
    ('C3', 'A3', 'B3'), # R17
    ('B3', 'A3', '~C70'), ('B75', 'A3', '~C70'), ('B3', 'A75', '~C70'), ('B75', 'A75', '~C70'), # R18
    ('B3', 'A3', '~C65'), ('B70', 'A3', '~C65'), ('B3', 'A70', '~C65'), ('B70', 'A70', '~C65'), # R19

    # clue 5
    ('~A2',), # R20
    ('A2', 'B2', 'C2'), # R21

    # constraints
    ('AF', 'AG', 'AH'), ('~AF', '~AG'), ('~AF', '~AH'), ('~AG', '~AH'), # R22
    ('BF', 'BG', 'BH'), ('~BF', '~BG'), ('~BF', '~BH'), ('~BG', '~BH'), # R23
    ('CF', 'CG', 'CH'), ('~CF', '~CG'), ('~CF', '~CH'), ('~CG', '~CH'), # R24

    ('A65', 'A70', 'A75'), ('~A65', '~A70'), ('~A65', '~A75'), ('~A70', '~A75'), # R25
    ('B65', 'B70', 'B75'), ('~B65', '~B70'), ('~B65', '~B75'), ('~B70', '~B75'), # R26
    ('C65', 'C70', 'C75'), ('~C65', '~C70'), ('~C65', '~C75'), ('~C70', '~C75'), # R27

    ('A2', 'A3', 'A4'), ('~A2', '~A3'), ('~A2', '~A4'), ('~A3', '~A4'), # R28
    ('B2', 'B3', 'B4'), ('~B2', '~B3'), ('~B2', '~B4'), ('~B3', '~B4'), # R29
    ('C2', 'C3', 'C4'), ('~C2', '~C3'), ('~C2', '~C4'), ('~C3', '~C4'), # R30

    ('AF', 'BF', 'CF'), ('~AF', '~BF'), ('~AF', '~CF'), ('~BF', '~CF'), # R31
    ('AG', 'BG', 'CG'), ('~AG', '~BG'), ('~AG', '~CG'), ('~BG', '~CG'), # R32
    ('AH', 'BH', 'CH'), ('~AH', '~BH'), ('~AH', '~CH'), ('~BH', '~CH'), # R33

    ('A65', 'B65', 'C65'), ('~A65', '~B65'), ('~A65', '~C65'), ('~B65', '~C65'), # R34
    ('A70', 'B70', 'C70'), ('~A70', '~B70'), ('~A70', '~C70'), ('~B70', '~C70'), # R35
    ('A75', 'B75', 'C75'), ('~A75', '~B75'), ('~A75', '~C75'), ('~B75', '~C75'), # R36

    ('A2', 'B2', 'C2'), ('~A2', '~B2'), ('~A2', '~C2'), ('~B2', '~C2'), # R37
    ('A3', 'B3', 'C3'), ('~A3', '~B3'), ('~A3', '~C3'), ('~B3', '~C3'), # R38
    ('A4', 'B4', 'C4'), ('~A4', '~B4'), ('~A4', '~C4'), ('~B4', '~C4'), # R39
]

In [ ]:
Symbols = (
    'AF','BF','CF',
    'AG','BG','CG',
    'AH','BH','CH',

    'A65','B65','C65',
    'A70','B70','C70',
    'A75','B75','C75',

    'A2','B2','C2',
    'A3','B3','C3',
    'A4','B4','C4',
)

Model =  {i:None for i in Symbols}

In [ ]:
def EvaluateKB(Clauses, Model):
    EvaluatedClauses = []

    for clause in Clauses:
        EvaluatedLiterals = []

        for literal in clause:
            if literal[0] == '~':
                if Model[literal[1:]] is None:
                    EvaluatedLiterals.append(None)
                else:
                    EvaluatedLiterals.append(not Model[literal[1:]])
            else:
                if Model[literal] is None:
                    EvaluatedLiterals.append(None)
                else:
                    EvaluatedLiterals.append(Model[literal])

        if True in EvaluatedLiterals:
            EvaluatedClauses.append(True)
        elif None in EvaluatedLiterals:
            EvaluatedClauses.append(None)
        else:
            EvaluatedClauses.append(False)

    if all(i is True for i in EvaluatedClauses):
        EvaluatedKB = True
    elif False in EvaluatedClauses:
        EvaluatedKB = False
    else:
        EvaluatedKB = None

    return EvaluatedKB

In [ ]:
import copy

def BackTrackingSearch(Clauses, Model, model_count=0):

    EvaluatedClauses = EvaluateKB(Clauses, Model)

    if EvaluatedClauses is True:
        return True, Model, model_count
    elif EvaluatedClauses is False:
        return False, None, model_count
    else:
        model_count += 1 # for 1(e) evaluation

        Model2 = copy.deepcopy(Model)
        NextSymbol = [i for i in Model2 if Model2[i] is None][0]

        Model2[NextSymbol] = True
        ModelTrue, SATModel, model_count = BackTrackingSearch(Clauses, Model2, model_count)
        if ModelTrue:
            return True, SATModel, model_count

        Model2[NextSymbol] = False
        ModelFalse, SATModel, model_count = BackTrackingSearch(Clauses, Model2, model_count)
        if ModelFalse:
            return True, SATModel, model_count

    return False, None, model_count

In [ ]:
result, SATModel, model_count = BackTrackingSearch(Clauses, Model)

print(result, SATModel)
print(f"Model Count: {model_count}")

True {'AF': False, 'BF': True, 'CF': False, 'AG': False, 'BG': False, 'CG': True, 'AH': True, 'BH': False, 'CH': False, 'A65': True, 'B65': False, 'C65': False, 'A70': False, 'B70': False, 'C70': True, 'A75': False, 'B75': True, 'C75': False, 'A2': False, 'B2': False, 'C2': True, 'A3': False, 'B3': True, 'C3': False, 'A4': True, 'B4': False, 'C4': False}
Model Count: 91


## Task (c)

Determine whether your KB is uniquely satisfiable.

In [ ]:
import itertools

Models = itertools.product((T,F), repeat=27)

TrueModelCount = 0
IterationCount = 0 # for 1(e) evaluation

for m in Models:
    IterationCount += 1
    TF, Result =  evaluateKB(m, KnowledgeBase)
    if TF:
        TrueModelCount += 1
        print(f"---\n True Model Count {TrueModelCount}: {TF}, {m}")
        print(Result)
        print(f"Iteration Count: {IterationCount}")

if TrueModelCount == 0:
    print("No true models found after iterating through all possible models.")

---
 True Model Count 1: True, (False, True, False, False, False, True, True, False, False, True, False, False, False, False, True, False, True, False, False, False, True, False, True, False, True, False, False)
[True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
Iteration Count: 97381292


## Task (d)

- Improve the backtracking search method by eliminating “trivial solutions”entailed by so-called unit clauses in the Conjunctive Normal Form of the KB that involve only one literal.
- Clauses such as (P,) or ('~P',) determine the state the literal must take for the KB to be satisfied.
- Write a function to identify such clauses and generate a Model in which the states of literals that can be determined from unit clauses are prepopulated.
- You do not need to modify the BackTrackingSearch function.

In [ ]:
def IdentifyUnitClauses(Clauses):
    UnitClauses = []

    for clause in Clauses:
        if len(clause) == 1:
            UnitClauses.append(clause)

    return UnitClauses

In [ ]:
UnitClauses = IdentifyUnitClauses(Clauses)

print(UnitClauses)

[('~CF',), ('~C75',), ('~BH',), ('~B2',), ('~C3',), ('~A2',)]


In [ ]:
def PrepopulateUnitClauses(Clauses, Model):
    UnitClauses = IdentifyUnitClauses(Clauses)

    for clause in UnitClauses:
        literal = clause[0]
        if literal[0] == '~':
            Model[literal[1:]] = False
        else:
            Model[literal] = True

    return Model

In [ ]:
PrepopulatedModel = PrepopulateUnitClauses(Clauses, Model)

print(PrepopulatedModel)

{'AF': None, 'BF': None, 'CF': False, 'AG': None, 'BG': None, 'CG': None, 'AH': None, 'BH': False, 'CH': None, 'A65': None, 'B65': None, 'C65': None, 'A70': None, 'B70': None, 'C70': None, 'A75': None, 'B75': None, 'C75': False, 'A2': False, 'B2': False, 'C2': None, 'A3': None, 'B3': None, 'C3': False, 'A4': None, 'B4': None, 'C4': None}


In [ ]:
BackTrackingSearch(Clauses, PrepopulatedModel)
result, SATModel, model_count = BackTrackingSearch(Clauses, PrepopulatedModel)

print(result, SATModel)
print(f"Model Count: {model_count}")

True {'AF': False, 'BF': True, 'CF': False, 'AG': False, 'BG': False, 'CG': True, 'AH': True, 'BH': False, 'CH': False, 'A65': True, 'B65': False, 'C65': False, 'A70': False, 'B70': False, 'C70': True, 'A75': False, 'B75': True, 'C75': False, 'A2': False, 'B2': False, 'C2': True, 'A3': False, 'B3': True, 'C3': False, 'A4': True, 'B4': False, 'C4': False}
Model Count: 67


## Task (e)

- For each of the methods of solution used in parts b, c, and d, calculate how many models were evaluated before a solution was found (you will need to
modify the code to do this).
- Does this correspond to your expectations?
- What factors might influence the relative performance of these methods on a specific problem?

**Solution: Students' Subject, Marks & their Graduation Date**

| Student | Subject | Mark | Graduated Date |
| ----------- | ----------- | ----------- | ----------- |
| Anjali | History | Very Good 65% | 4th July |
| Brian | French | Outstanding 75% | 3rd July |
| Chen | Geography | Excellent 70% | 2nd July |

**Unique SAT Model**

model =  [F, T, F, F, F, T, T, F, F, T, F, F, F, F, T, F, T, F, F, F, T, F, T, F, T, F, F]

For this project, three approaches were applied to find the solution.
<br>

The first approach 1(b) is a backtracking search which is a SAT solver. It looks for a model that satisfies the Knowledge Base (KB). To apply backtracking search to the KB, sentences were first transformed into the Conjuctive Normal Form (CNF).
<br>

The second approach 1(c) is a model checking, which uses a crude way of enumerating and iterating through the truth table to find all the possible models that satisfy the KB.
<br>

The third approach 1(d) is a improved version of backtracking search. It trims the search space by identifying the unit clauses and prepopulating the model with their states.
<br>

Among the three approaches, 1(d) is the most efficient approach which found the soulation with 67 iterations, followed by 1(b) with 91 iterations. On the other hand, 1(c) has to iterate 97,381,292 times to get the unique SAT model. 1(d) & 1(b) performed their tasks instantly while 1(c) took approximately 36 minutes and 22 seconds to iterate all 134,217,728 possible models in the truth table.
<br>

1(d) and 1(b) are reliable approaches compared to 1(c). However, these models stopped after they found their first SAT model which made it difficult to determine the uniqueness of the SAT model.  Another problem is transparency. Due to this, it is difficult to improve the KB with just backtracking search.
<br>

On the other hand, 1(c), model checking is reliable in tracking errors or improvment. But the downside of it is consumption of time and computational power. These increase in relation to the number of atomic propositions. The proposed model has 27 atomic propositions which leads to over 130 million possible models.
<br>

Other problems faced during the development of this KB are balancing the complexity & interpretability of the sentences. Assigning the right amount of atomic propositions was also challenging. One last problem is transforming a sentence into CNF. As an example, R5 is very complex to transform into CNF manually.
<br>

In conclusion, back tracking search with trimming is the most reliable approach. However, to imporve the model performance, model checking was required.